### DISCUSSION TITLE
Patching custom font file?

### DISCUSSION THREAD

**Discussion by rodhash** [ryanoasis/nerd-fonts#1732] [General]
I have a custom font file `my-font.ttf`, how do I add the nice glyphs from it into my `CodeNewRomanNerdFont-Regular.otf` font?

Tried a few different things already for instance moving this `my-font.ttf` into `src/glyphs` and issued the fontforge cmd but my glyphs were not added to `CodeNewRomanNerdFont-Regular.otf`:

```sh
 $ fontforge -script ./font-patcher --name MyCustomFont toPatch/CodeNewRomanNerdFont-Regular.otf -out result
```
ps: I also tried using `-c` or `--complete`. Tried many other flags too.

<img width="642" alt="image" src="https://github.com/user-attachments/assets/d162adc6-1171-4c49-b5ad-6378acf7e04e">

Is that the proper place for it? Keep it under `src/glyphs`?

I'm able to install this `my-font.ttf` into my Mac and my `iTerm2` is able to use it just fine. Here I can see the new glyphs in my terminal so the font seems to work fine.

But my intention here is patching  my `CodeNewRomanNerdFont-Regular.otf` font in order to add the new glyphs to it.

Suggestions?

Also tried merging both fonts as in this [link](https://superuser.com/a/1594984/1120770) but doesn't seem to work. I mean a new font is generated but I don't see the new glyphs added.

Thanks.

**Comment by Finii** [comment]
I believe you
* Have already a Nerd Font patched font (i.e. CodeNewRoman NF)
* Want to add some special 'your own' glyphs
* That glyphs are in `my-font.ttf`

So, 'take all glyphs in `my-font.ttf` and add them to `CodeNewRomanNF`.

In your command line above you do not mention `my-font.ttf`:

> ```console
> $ fontforge -script ./font-patcher --name MyCustomFont toPatch/CodeNewRomanNerdFont-Regular.otf -out result
> ```

The `--help` says:

![image](https://github.com/user-attachments/assets/d27a665e-34aa-46ae-991c-58dcdd82c40f)

So maybe you need


```console
$ fontforge -script ./font-patcher --custom my-font.ttf --name MyCustomFont toPatch/CodeNewRomanNerdFont-Regular.otf -out result
```

You do not need to put `my-font.ttf` into any special directory.

The glyph scaling/shifting might be unsatisfactory, as the patcher does not 'know' much about the glyphs it shall add; some default is used that might or might not be ok for you,

[Here](https://github.com/ryanoasis/nerd-fonts/wiki/ScriptOptions) is a slightly more verbose help ;-)

![image](https://github.com/user-attachments/assets/b514b9a6-8433-4ea9-b9c0-5643cecc2f50)

**Reply by rodhash** [reply]
Oh boy I missed this "custom" parameter now my custom glyphs were added! :)

`The glyph scaling/shifting might be unsatisfactory, as the patcher does not 'know' much about the glyphs it shall add; some default is used that might or might not be ok for you`

Yeah it seems something is off .. I see the new glyphs but they are not properly aligned.

**Reply by rodhash** [reply]
What would be these default? I think I might need to play around with those values to make my glyphs align properly but I'm not finding the proper flags in the docs.

**Reply by Finii** [reply]
There are no commandline options, change the defaults in the script here:

```python
        CUSTOM_ATTR = {
            # previous custom scaling => do not touch the icons
            # 'default': {'align': 'c', 'valign': '', 'stretch': '', 'params': {}}
            'default': {'align': 'c', 'valign': 'c', 'stretch': 'pa', 'params': {'careful': self.args.careful}}
        }
```
_`font-patcher` around line 950_

Find possible values further up in the script. There is also this explanation of the flags:

```python
        # Stretch 'xz' or 'pa' (preserve aspect ratio)
        # Supported params: overlap | careful | xy-ratio | dont_copy | ypadding
        # Overlap value is used horizontally but vertically limited to 0.01
        # Careful does not overwrite/modify existing glyphs
        # The xy-ratio limits the x-scale for a given y-scale to make the ratio <= this value (to prevent over-wide glyphs)
        # '1' means occupu 1 cell (default for 'xy')
        # '2' means occupy 2 cells (default for 'pa')
        # '!' means do the 'pa' scaling even with non mono fonts (else it just scales down, never up)
        # '^' means that scaling shall fill the whole cell and not only the icon-cap-height (for mono fonts, other always use the whole cell)
        # Dont_copy does not overwrite existing glyphs but rescales the preexisting ones
        #
        # Be careful, stretch may not change within a ScaleRule!
```

You can also use different rules for different custom glyphs, just look how that is done in the other cases.

**Reply by rodhash** [reply]
Oh it's right there, thanks again! Later after work will play with those values.

**Comment by Finii** [comment]
:+1: for finding out the fix for your initial `CRITICAL: Can not detect sane font width` problem

**Comment by BEST8OY** [comment]
@Finii 
FontForge auto width option is really nice though for shifting issue!

**Reply by Finii** [reply]
Not sure what you mean.

**Reply by BEST8OY** [reply]
> Not sure what you mean.

So here is the thing
I tried these two commands
1.```fontforge --script ./font-patcher --name CustomFont --variable-width-glyphs --custom /home/hl/font/xxx.ttf ../../font/CaskaydiaCoveNerdFontPropo-Regular.ttf ```

2.```fontforge --script ./font-patcher --name CustomFont --custom /home/hl/font/xxx.ttf ../../font/CaskaydiaCoveNerdFontPropo-Regular.ttf```

The problem with the first one is that it causes some apps to crash when used (Linux) (I know for sure `sublime` is crashing because of it)

The problem with the second one is that it causes the added glyphs to be shifted a bit to the right (width shifting issue) and the solution that worked out for me was to use FontForge Metric > Auto Width to fix the width for the newly added glyphs. 
I'm not sure if there is a smarter way to go around this!

**Reply by Finii** [reply]
Well, the scaling/shifting settings is in `CUSTOM_ATTR` as detailed above.

Depending on which shift is unwanted, remove that `c`, and or also the `pa`.

**Reply by BEST8OY** [reply]
> Well, the scaling/shifting settings is in `CUSTOM_ATTR` as detailed above.

I figured that these options are not my problem! (perhaps) (I get better result with these default options)
My problem is that I want the cell for the added glyphs to scale to the size of the glyph, but it stays at 1200
**Is there an option for this?**

For example this is the current result
![image](https://github.com/user-attachments/assets/40103104-cd94-4ee5-8fde-3335de0af7ff)
I want the final result be like this
![image](https://github.com/user-attachments/assets/909bf67e-1535-455d-98d6-330afa87e9fa)

**Reply by Finii** [reply]
This is what `--variable-width-glyphs` does. Do not understand why the resultant font should crash an application 🤔

**Reply by BEST8OY** [reply]
Is this situation in the scope of --variable-width-glyphs behavior?
It's as if it doesn't recognize spaces
1.With this option
Look at the timer Icon
![image](https://github.com/user-attachments/assets/244e58c2-120e-42cb-a99d-4234cf38803c)

2.Without this option
![image](https://github.com/user-attachments/assets/56afd97d-a42b-4c39-8864-daee2e70a870)

**Reply by Finii** [reply]
Ah yes, it removes bearings (borders).
Well, then I guess there is no better way to patch the font as you do already.
`--custom` is not very flexible and just a byproduct.

For glyphs that need to have space around (bearings) in the normal sets I define a negative `"overlap":` value.
So if the amount of these buffer zones around the bounding box of the added glyphs shall always be the same, you can add that to the 

```
'default': {'align': 'c', 'valign': 'c', 'stretch': 'pa', 'params': {'overlap': -0.1}}
```

Ten percent (10%) border in total, i.e. 5% width border left and 5% right.

**Reply by Finii** [reply]
Hmm, but here is also no border and you said you want it to look like this

![image](https://github.com/user-attachments/assets/909bf67e-1535-455d-98d6-330afa87e9fa)

**Reply by BEST8OY** [reply]
The overlap option when using --variable-width-glyphs fixes `sublime` crash.

But since I still need my spaces I go with my way (fixing the width manually with FontForge `Auto Width` option)
